pkg/ipamd/ipamd_test.go

```diff
@@ -745,6 +746,41 @@ func testIncreasePrefixPool(t *testing.T, useENIConfig bool) {
 	mockContext.increaseDatastorePool(ctx)
 }
 
+func TestDecreaseIPPool(t *testing.T) {
+	m := setup(t)
+	defer m.ctrl.Finish()
+
+	mockContext := &IPAMContext{
+		awsClient:          m.awsutils,
+		warmIPTarget:       1,
+		lastDecreaseIPPool: time.Now().Add(-60 * time.Second),
+	}
+	mockContext.reconcileCooldownCache.cache = make(map[string]time.Time)
+
+	testAddr1 := net.IPNet{IP: net.ParseIP(ipaddr01), Mask: net.IPv4Mask(255, 255, 255, 255)}
+	testAddr2 := net.IPNet{IP: net.ParseIP(ipaddr02), Mask: net.IPv4Mask(255, 255, 255, 255)}
+	testAddr11 := net.IPNet{IP: net.ParseIP(ipaddr11), Mask: net.IPv4Mask(255, 255, 255, 255)}
+	testAddr12 := net.IPNet{IP: net.ParseIP(ipaddr12), Mask: net.IPv4Mask(255, 255, 255, 255)}
+
+	mockContext.dataStore = testDatastore()
+
+	mockContext.dataStore.AddENI(primaryENIid, primaryDevice, true, false, false)
+	mockContext.dataStore.AddIPv4CidrToStore(primaryENIid, testAddr1, false)
+	mockContext.dataStore.AddIPv4CidrToStore(primaryENIid, testAddr2, false)
+	mockContext.dataStore.AssignPodIPv4Address(datastore.IPAMKey{ContainerID: "container1"}, datastore.IPAMMetadata{K8SPodName: "pod1"})
+
+	mockContext.dataStore.AddENI(secENIid, secDevice, true, false, false)
+	mockContext.dataStore.AddIPv4CidrToStore(secENIid, testAddr11, false)
+	mockContext.dataStore.AddIPv4CidrToStore(secENIid, testAddr12, false)
+	mockContext.dataStore.AssignPodIPv4Address(datastore.IPAMKey{ContainerID: "container2"}, datastore.IPAMMetadata{K8SPodName: "pod2"})
+
+	m.awsutils.EXPECT().DeallocPrefixAddresses(gomock.Any(), gomock.Any()).Times(1)
+	m.awsutils.EXPECT().DeallocIPAddresses(gomock.Any(), gomock.Any()).Times(1)
+
+	mockContext.decreaseDatastorePool(10 * time.Second)
+	mockContext.decreaseDatastorePool(10 * time.Second)
```

Comment:
why are we calling this twice?

Reply:
I added the call just to ensure that multiple deallocations back to back does not take the target below warm ip. If it is extraneous, i can drop it. I added a comment.

Reply:
Ah, I see, that's fine with me. I think we should assert the datastore target state (calling `datastoreTargetState`) is fine, before the first call to `decreaseDatastorePool`, and then after each. That way, we can assert that before we do anything, we are two over, and after the first call, we are 0 over, and after the second call, we are still 0 over

Reply:
I updated the assertions